### PR TITLE
[Merged by Bors] - feat: `LawfulOrd` and `LinearOrder`

### PR DESCRIPTION
**Pull request by digama0**
- [x] depends on: #12068
- [x] depends on: leanprover/std4#730

This adds the necessary instances for types from Mathlib to play well with the new classes from Std. This solves one of the `mathlibSorry`'s from the [EmptyHexagon project](https://github.com/bsubercaseaux/EmptyHexagonLean/blob/98dee70a01acc111977bc8f9c369c59a922221d7/Lean/Geo/ToMathlib.lean#L217-L219).

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by urkud** [comment]
There are multiple merge conflicts. Are you going to come back to this PR?

File: Mathlib/Data/Prod/Lex.lean [hunk 1]
```diff
@@ -39,6 +39,12 @@ namespace Prod.Lex
 -- This will be fixed in nightly-2022-11-30
 @[inherit_doc] notation:35 α " ×ₗ " β:34 => _root_.Lex (Prod α β)
 
+instance (α β : Type*) [BEq α] [BEq β] : BEq (α ×ₗ β) where
+  beq a b := ofLex a == ofLex b
```
Comment (urkud) [inline Mathlib/Data/Prod/Lex.lean:43]: This should be done for any `Lex _`.

File: Mathlib/Data/Prod/Lex.lean [hunk 2]
```diff
@@ -133,18 +139,37 @@ instance partialOrder (α β : Type*) [PartialOrder α] [PartialOrder β] : Part
       exact @antisymm _ (Prod.Lex _ _) _ }
 #align prod.lex.partial_order Prod.Lex.partialOrder
 
+instance instOrdLexProd [Ord α] [Ord β] : Ord (α ×ₗ β) where
+  compare := compareLex (compareOn fun x => (ofLex x).1) (compareOn fun x => (ofLex x).2)
```
Comment (urkud) [inline Mathlib/Data/Prod/Lex.lean:143]: There is `def lexOrd` in core. Should we reuse it here?

File: Mathlib/Data/Prod/Lex.lean [hunk 3]
```diff
@@ -124,16 +120,36 @@ instance partialOrder (α β : Type*) [PartialOrder α] [PartialOrder β] : Part
...
+    decidableLT := Prod.Lex.decidable _ _
+    decidableEq := instDecidableEqLex _
+    compare_eq_compareOfLessAndEq := fun a b => by
+      have : DecidableRel (· < · : α ×ₗ β → α ×ₗ β → Prop) := Prod.Lex.decidable _ _
+      have : BEqOrd (α ×ₗ β) := ⟨by
+        simp [compare_def, compareLex, compareOn, Ordering.then_eq_eq, compare_eq_iff_eq]⟩
+      have : LTOrd (α ×ₗ β) := ⟨by
+        simp [compare_def, compareLex, compareOn, Ordering.then_eq_lt, lt_iff,
+          compare_lt_iff_lt, compare_eq_iff_eq]⟩
+      convert LTCmp.eq_compareOfLessAndEq (cmp := compare) a b }
```
Comment (urkud) [inline Mathlib/Data/Prod/Lex.lean:152]: dedent by 2 spaces?

File: Mathlib/Data/Prod/Lex.lean [hunk 4]
```diff
@@ -124,16 +120,36 @@ instance partialOrder (α β : Type*) [PartialOrder α] [PartialOrder β] : Part
...
-    decidableLT := Prod.Lex.decidable _ _,
-    decidableEq := Lex.decidableEq _ _, }
-
-instance [Ord α] [Ord β] : Ord (α ×ₗ β) where
-  compare := compareLex (compareOn (·.1)) (compareOn (·.2))
+    le_total := total_of (Prod.Lex _ _)
+    decidableLE := Prod.Lex.decidable _ _
+    decidableLT := Prod.Lex.decidable _ _
+    decidableEq := instDecidableEqLex _
+    compare_eq_compareOfLessAndEq := fun a b => by
```
Comment (urkud) [inline Mathlib/Data/Prod/Lex.lean:145]: Should we have a constructor from some *Ord typeclasses?

**Comment by urkud** [comment]
Thanks!
bors d+

File: Mathlib/Data/Prod/Lex.lean (hunk 3 above)
Comment (digama0) [reply Mathlib/Data/Prod/Lex.lean:152]: I don't see why? Perhaps you are misreading, all the indents here appear necessary.

File: Mathlib/Data/Prod/Lex.lean (hunk 4 above)
Comment (digama0) [reply Mathlib/Data/Prod/Lex.lean:145]: Probably, but it's not clear to me whether there is an exploitable pattern here.

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**